Draupnir proxy

Chasethechicken opened the pull request:
Mjolnir / Draupnir can receive abuse reports in the management room.

There are two options to configure this:
1. Poll reports from the admin-API
2. Redirect the endpoint to the draupnir container using a reverse proxy

The second option seems to be the preferred one.

This PR configures the MDAD-managed traefik to proxy the relevant requests to Draupnir (the second option). To enable it set `matrix_bot_draupnir_abuse_reporting_enabled: true`.

If this gets merged I'd be willing to create a PR adding the same options for Mjolnir, which should more or less be a big find and replace.

spantaleev commented on roles/custom/matrix-bot-draupnir/templates/production.yaml.j2:
```diff
@@ -261,4 +258,4 @@ pollReports: false
 # Whether or not new reports, received either by webapi or polling,
 # should be printed to our managementRoom.
-displayReports: false
+displayReports: {{ matrix_bot_draupnir_abuse_reporting_enabled }}
```
It's probably better to add a new variable for this. Its default value can be defined in terms of `matrix_bot_draupnir_abuse_reporting_enabled`.

Based the setting key, the variable should be named `matrix_bot_draupnir_display_reports`.

Perhaps `abuseReporting.enabled` should also receive its own variable (or use `matrix_bot_draupnir_abuse_reporting_enabled`).

Then, `web.enabled` can be its own variable too, with a default defined in terms of `matrix_bot_draupnir_abuse_reporting_enabled`:

```yml
matrix_bot_draupnir_web_enabled: "{{ matrix_bot_draupnir_abuse_reporting_enabled }}"
```

`matrix_bot_draupnir_abuse_reporting_enabled` should probably be renamed to `matrix_bot_draupnir_web_abuse_reporting_enabled`.

All these variables names are meant to indicate where the variable is being used. We usually try to make variables take their names from the setting path that they control.

-----

All this would let people:

- enable the web-interface manually (for other purposes), without enabling abuse-reporting
- control `displayReports` separately (being able to turn it off, even if abuse reporting is enabled)

spantaleev commented:
Thank you for working on this feature!

I've posted some feedback to improve consistency and simplify some things.

Chasethechicken commented on docs/configuring-playbook-bot-draupnir.md:
````diff
@@ -146,7 +146,10 @@ matrix_bot_draupnir_configuration_extension_yaml: |
-While this playbook uses reverse proxies, it does not yet implement this.
+If you are using traefik, this playbook can set this up for you:
+```yaml
+matrix_bot_draupnir_web_abuse_reporting_enabled: true
````
Should the default way of enabling this feature be `matrix_bot_draupnir_web_abuse_reporting_enabled` or `matrix_bot_draupnir_abuse_reporting_enabled`?

The first option requires less variables and thus less complexity.

The second option seems more intuitive, since from a user perspective this has nothing to do with web, the abuse reports are just sent as messages to the moderation room.

spantaleev replied:
I see your point! I guess keeping it `matrix_bot_draupnir_abuse_reporting_enabled` makes sense too. Feel free to do it that way

Chasethechicken commented:
Sorry about taking so long, things got in the way and I kind of forgot about this.
This is my first contribution to MDAD, so your feedback is very much appreciated.

I think this should address all of your points, let me know if there is still something left to improve.

Also, do you want me to rebase this into a few nicer commits or are you going to squash this into one commit anyway?

spantaleev commented:
Sure! Thanks for soldiering on and perfecting this!
I can squash-merge when ready, so it's up to you if you'd like to clean up by rebasing explicitly or not.

Chasethechicken commented:
I have renamed the variable back to `matrix_bot_draupnir_abuse_reporting_enabled`.
I have also rebased this into 2 commits, so I think this is ready now.

spantaleev commented on roles/custom/matrix-bot-draupnir/defaults/main.yml:
```diff
@@ -96,3 +105,25 @@ matrix_bot_draupnir_configuration_extension: "{{ matrix_bot_draupnir_configurati
+matrix_bot_draupnir_container_labels_traefik_enabled: false
+matrix_bot_draupnir_container_labels_traefik_docker_network: "{{ matrix_draupnir_bot_container_network }}"
+matrix_bot_draupnir_container_labels_traefik_hostname: "{{ matrix_synapse_container_labels_traefik_hostname }}"
+matrix_bot_draupnir_container_labels_traefik_path_prefix: "^/_matrix/client/(r0|v3)/rooms/([^/]*)/report/"
```
Maybe it's better to rename this to `matrix_bot_draupnir_container_labels_traefik_path_regexp`, if it's going to be using regular expressions and will be passed to `PathRegexp`.

We've done the same in `matrix-media-repo`. See [here](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/33d5b0d991473d69c0a302b8684225798b247b94/roles/custom/matrix-media-repo/defaults/main.yml#L84-L85) and [here](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/33d5b0d991473d69c0a302b8684225798b247b94/roles/custom/matrix-media-repo/tasks/validate_config.yml#L32).

spantaleev commented on roles/custom/matrix-bot-draupnir/tasks/setup_install.yml:
```diff
@@ -17,6 +17,17 @@
     - {path: "{{ matrix_bot_draupnir_docker_src_files_path }}", when: "{{ matrix_bot_draupnir_container_image_self_build }}"}
   when: "item.when | bool"
 
+- name: Ensure matrix-bot-draupnir is configured
```
```suggestion
- name: Ensure matrix-bot-draupnir support files installed
```

spantaleev commented on roles/custom/matrix-bot-draupnir/templates/production.yaml.j2:
```diff
@@ -227,32 +227,29 @@ health:
+  # to configure a reverse proxy, see e.g. test/nginx.conf
+  abuseReporting:
+    # Whether to enable this feature.
+    enabled: {{ matrix_bot_draupnir_abuse_reporting_enabled }}
```
```suggestion
    enabled: {{ matrix_bot_draupnir_abuse_reporting_enabled | to_json }}
```

spantaleev commented on roles/custom/matrix-bot-draupnir/templates/production.yaml.j2:
```diff
@@ -261,4 +258,4 @@ pollReports: false
 # Whether or not new reports, received either by webapi or polling,
 # should be printed to our managementRoom.
-displayReports: false
+displayReports: {{ matrix_bot_draupnir_display_reports }}
```
```suggestion
displayReports: {{ matrix_bot_draupnir_display_reports | to_json }}
```

Chasethechicken commented:
@spantaleev I implemented all your suggestions. Is there still something I can improve?
If you just did not get around to reviewing this yet, then no worries.
